servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/SequentialExecutor.java

```diff
@@ -0,0 +1,111 @@
+
+import java.util.concurrent.Executor;
+import java.util.concurrent.atomic.AtomicReference;
+import javax.annotation.Nullable;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * A concurrency primitive for providing thread safety without using locks.
+ *
+ * A {@link SequentialExecutor} is queue of tasks that are executed one at a time in the order they were
+ * received. This provides a way to serialize work between threads without needing to use locks which can
+ * result in thread contention and thread deadlock scenarios.
+ */
+final class SequentialExecutor implements Executor {
+
+    /**
+     * Handler of exceptions thrown by submitted Runnables.
+     */
+    @FunctionalInterface
+    public interface ExceptionHandler {
+
+        /**
+         * Handle the exception thrown from a submitted Runnable.
+         * Note that if this method throws the behavior is undefined.
+         *
+         * @param ex the Throwable thrown by the Runnable.
+         */
+        void onException(Throwable ex);
+    }
+
+    private final ExceptionHandler exceptionHandler;
+    private final AtomicReference<Cell> tail = new AtomicReference<>();
+
+    SequentialExecutor(final ExceptionHandler exceptionHandler) {
+        this.exceptionHandler = requireNonNull(exceptionHandler, "exceptionHandler");
+    }
+
+    @Override
+    public void execute(Runnable command) {
+        // Make sure we propagate any sync contexts.
+        command = AsyncContext.wrapRunnable(requireNonNull(command, "command"));
+        final Cell next = new Cell(command);
+        Cell t = tail.getAndSet(next);
+        if (t != null) {
+            // Execution already started. Link the old tail to the new tail.
+            t.next = next;
+        } else {
+            // We are the first element in the queue so it's our responsibility to drain.
+            // Note that the getAndSet establishes the happens before with relation to the previous draining
+            // threads since we must successfully perform a CAS operation to terminate draining.
+            drain(next);
+        }
+    }
+
+    private void drain(Cell next) {
+        for (;;) {
+            assert next != null;
+            try {
+                next.runnable.run();
+            } catch (Throwable ex) {
+                exceptionHandler.onException(ex);
+            }
+
+            // Attempt to get the next element.
+            Cell n = next.next;
+            if (n == null) {
+                // There doesn't seem to be another element linked. See if it was the tail and if so terminate draining.
+                // Note that a successful CAS established a happens-before relationship with future draining threads.
+                if (tail.compareAndSet(next, null)) {
+                    break;
+                }
+                // next isn't the tail but the link hasn't resolved: we must poll until it does.
+                while ((n = next.next) == null) {
```

Comment:
nit: personal taste, but i think spinning on a CAS while you set the tail, might be easier to read / follow through.

Reply:
We talked offline. The flow of CAS'ing the tail is race prone on setting the linkage from the old tail to the new. If you try to do the linkage before the CAS then racing submissions can corrupt the link and if you do it after it turns into essentially the same flow.